File: src/graph/schema.py
from enum import Enum
from typing import List
from pydantic import BaseModel, Field

class MathEntityType(str, Enum):
    AXIOM = "Axiom"
    DEFINITION = "Definition"
    LEMMA = "Lemma"
    THEOREM = "Theorem"
    COROLLARY = "Corollary"
    PROOF = "Proof"
    EXAMPLE = "Example"
    CONCEPT = "Concept"

class MathRelationType(str, Enum):
    USES_AXIOM = "USES_AXIOM"
    USES_DEFINITION = "USES_DEFINITION"
    USES_LEMMA = "USES_LEMMA"
    PROVES = "PROVES"
    COROLLARY_OF = "COROLLARY_OF"
    PREREQUISITE_FOR = "PREREQUISITE_FOR"
    DEPENDS_ON = "DEPENDS_ON"

class GraphNode(BaseModel):
    name: str = Field(..., description="Name of the mathematical concept, theorem, definition, formula, etc.")
    entity_type: MathEntityType = Field(..., description="The type of mathematical entity")
    description: str = Field("", description="Short summary or definition of the entity")

class GraphEdge(BaseModel):
    source: str = Field(..., description="Name of the source node entity")
    target: str = Field(..., description="Name of the target node entity")
    relation: MathRelationType = Field(..., description="Directional mathematical relationship between source and target")

class MathEntityExtraction(BaseModel):
    nodes: List[GraphNode] = Field(default_factory=list, description="List of mathematical entities extracted from text")
    edges: List[GraphEdge] = Field(default_factory=list, description="List of relationships between extracted entities")

ALLOWED_ENTITIES: List[str] = [e.value for e in MathEntityType]
ALLOWED_RELATIONS: List[str] = [r.value for r in MathRelationType]

SCHEMA_SYSTEM_PROMPT = """
You are a mathematical knowledge graph extraction engine.
Extract mathematical entities (Concepts, Theorems, Definitions, Proofs, Formulas, Examples) and their directional relationships from the text.
Only extract clear, relevant mathematical entities.
"""

File: go_backend/graph/schema.go
package graph

type MathEntityType string

const (
	Axiom      MathEntityType = "Axiom"
	Definition MathEntityType = "Definition"
	Lemma      MathEntityType = "Lemma"
	Theorem    MathEntityType = "Theorem"
	Corollary  MathEntityType = "Corollary"
	Proof      MathEntityType = "Proof"
	Formula    MathEntityType = "Formula"
	Example    MathEntityType = "Example"
	Concept    MathEntityType = "Concept"
)

type MathRelationType string

const (
	UsesAxiom       MathRelationType = "USES_AXIOM"
	UsesDefinition  MathRelationType = "USES_DEFINITION"
	UsesLemma       MathRelationType = "USES_LEMMA"
	Proves          MathRelationType = "PROVES"
	CorollaryOf     MathRelationType = "COROLLARY_OF"
	PrerequisiteFor MathRelationType = "PREREQUISITE_FOR"
	DependsOn       MathRelationType = "DEPENDS_ON"
	Contains        MathRelationType = "CONTAINS"
)

type Provenance struct {
	DocID          string `json:"doc_id"`
	DocTitle       string `json:"doc_title"`
	DocPath        string `json:"doc_path,omitempty"`
	PageNumber     int    `json:"page_number,omitempty"`
	SectionHeading string `json:"section_heading,omitempty"`
	ExactQuote     string `json:"exact_quote,omitempty"`
}

type ConceptTaxonomy struct {
	Domain    string `json:"domain"`
	Subdomain string `json:"subdomain"`
	Topic     string `json:"topic"`
}

type GraphNode struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	EntityType  MathEntityType  `json:"entity_type"`
	Taxonomy    ConceptTaxonomy `json:"taxonomy"`
	Aliases     []string        `json:"aliases,omitempty"`
	Description string          `json:"description"`
	Provenance  []Provenance    `json:"provenance,omitempty"`
}

type GraphEdge struct {
	Source      string           `json:"source"`
	Target      string           `json:"target"`
	Relation    MathRelationType `json:"relation"`
	Description string           `json:"description,omitempty"`
	Provenance  []Provenance     `json:"provenance,omitempty"`
}

type MathEntityExtraction struct {
	Nodes []GraphNode `json:"nodes"`
	Edges []GraphEdge `json:"edges"`
}
